Bottom: 8df18e258b71d76573609f28298400e9c9cd6a00
Top:    5e455342714ec1d87941b7c101230419352aefad
Author: Alexandre Oliva <aoliva@redhat.com>
Date:   2016-03-02 01:21:07 -0300

Refresh of char-type.patch

---

diff --git a/include/gcc-c-fe.def b/include/gcc-c-fe.def
index 25ace1c..25f83f9 100644
--- a/include/gcc-c-fe.def
+++ b/include/gcc-c-fe.def
@@ -125,7 +125,8 @@ GCC_METHOD3 (gcc_type, build_function_type,
 	     const struct gcc_type_array *, /* Argument ARGUMENT_TYPES.  */
 	     int /* bool */)               /* Argument IS_VARARGS.  */
 
-/* Return an integer type with the given properties.  */
+/* Return an integer type with the given properties.
+   Pass zero in IS_UNSIGNED and SIZE_IN_BYTES to get type 'char'.  */
 
 GCC_METHOD2 (gcc_type, int_type,
 	     int /* bool */,               /* Argument IS_UNSIGNED.  */
diff --git a/libcc1/libcc1plugin.cc b/libcc1/libcc1plugin.cc
index 6f7b527..abcf63a 100644
--- a/libcc1/libcc1plugin.cc
+++ b/libcc1/libcc1plugin.cc
@@ -698,8 +698,13 @@ gcc_type
 plugin_int_type (cc1_plugin::connection *self,
 		 int is_unsigned, unsigned long size_in_bytes)
 {
-  tree result = c_common_type_for_size (BITS_PER_UNIT * size_in_bytes,
-					is_unsigned);
+  tree result;
+
+  if (!size_in_bytes && !is_unsigned)
+    result = char_type_node;
+  else
+    result = c_common_type_for_size (BITS_PER_UNIT * size_in_bytes,
+				     is_unsigned);
   if (result == NULL_TREE)
     result = error_mark_node;
   else
